torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py

```diff
@@ -278,52 +311,88 @@ def _annotate_conv2d(
-            return
-
-        # bias and output act
-        addmm_node.meta["target_dtype_info"] = {
-            "bias_obs_or_fq_ctr": get_bias_obs_or_fq_ctr(quantization_config),
-            "input_act_obs_or_fq_ctr": PlaceholderObserver.with_args(dtype=torch.float),
-            "output_act_obs_or_fq_ctr": get_act_obs_or_fq_ctr(quantization_config),
-            "bias_index": 0,
-            "_annotated": True,
-        }
-        # input act
-        view_node.meta["target_dtype_info"] = {
-            "input_act_obs_or_fq_ctr": get_act_obs_or_fq_ctr(quantization_config),
-            "output_act_obs_or_fq_ctr": PlaceholderObserver.with_args(
-                dtype=torch.float
-            ),
-            "_annotated": True,
-        }
-        # weight
-        t_node.meta["target_dtype_info"] = {
-            "input_act_obs_or_fq_ctr": get_weight_obs_or_fq_ctr(quantization_config),
-            "output_act_obs_or_fq_ctr": PlaceholderObserver.with_args(
-                dtype=torch.float
-            ),
-            "_annotated": True,
-        }
+        graph = gm.graph
+        patterns = []
+        """
+        Annotate linear nodes:
+        This is done by tracing linear patterns for various input shapes that give
+        distinct pattern graph.
+        Order matters here since without that we get overlapping matches, resulting
+        in wrong annotations.
+        We will really plan to move this as graph matching utils supported by compiler/core team.
+        More details can be found here: (put doc link)
+        """
+        patterns.extend(_get_linear_patterns([8, 8, 8, 8]))
+        patterns.extend(_get_linear_patterns([8, 8, 8]))
+        patterns.extend(_get_linear_patterns([8, 8]))
+        matches: List[InternalMatch] = []
+        for pattern in patterns:
+            subgraph_matcher = SubgraphMatcher(pattern, ignore_literals=True)
+            matches.extend(subgraph_matcher.match(graph))
+        for match in matches:
+            weight_or_bias = []
+            act_node = None
+            if len(match.returning_nodes) != 1:
+                raise ValueError("Linear pattern must have only one returning node")
+            output_node = match.returning_nodes[0]
+            for ph in match.placeholder_nodes:
+                if ph.op == "get_attr":
+                    weight_or_bias.append(ph)
+                else:
+                    act_node = ph
+            weight_node = None
+            bias_node = None
+            for ph in weight_or_bias:
+                weight_or_bias = getattr(gm, ph.target)
+                if weight_or_bias.ndim == 2:
+                    weight_node = ph
+                if weight_or_bias.ndim == 1:
+                    bias_node = ph
+
+            # bias and output act
+            if _is_annotated([act_node]) == False:
+                act_node.meta["target_dtype_info"] = {
+                    "input_act_obs_or_fq_ctr": None,
```

Comment:
May I ask a question why `input_act_obs_or_fq_ctr` is None here for `act_node`? Does it mean we don't insert observer for the input of  `act_node`?

Reply:
`act_node` here is actually the node that produces the activation for linear, so the output of `act_node` is the input for linear